modules/minimp3/audio_stream_mp3.cpp

```diff
@@ -32,7 +32,6 @@
 #define MINIMP3_FLOAT_OUTPUT
 #define MINIMP3_IMPLEMENTATION
 #define MINIMP3_NO_STDIO
-
 #include "audio_stream_mp3.h"
```

Comment:
Why remove the empty line here? And why are these defines above includes at all?

Reply:
Those defines are needed for the single-header `minimp3_ex.h`, they need to be set before including that header for the first time.

They should only be defined once, but the header is included in multiple locations, so the code is structured so that this .cpp is responsible for the implementation (`minimp3_ex.h` is included in `audio_stream_mp3.h`), while other Godot files which include `audio_stream_mp3.h` (thus `minimp3_ex.h` again) will only use the actual header part of `minimp3_ex.h`.

Removing the space before the include helps clarify that those defines directly impact the include, and thus shouldn't be moved.

Yes, single header libraries are dumb.

An alternative could be to add a `thirdparty/minimp3/minimp3_ex.cpp` that includes the header and set the implementation flags, so we don't need to do it in module code.

Reply:
Thanks, suspected as much.

> They should only be defined once, but the header is included in multiple locations

Isn't that what all those guards in headers are for? To avoid duplicate definitions when the file is included multiple times? In other words, shouldn't it be safe to add these defines in the header after `#define AUDIO_STREAM_MP3_H`? Although, it would still be awkwardly positioned between includes. But at least directly next to what requires it?

Reply:
That's where single header libraries are a heresy. The header guard only covers the "header" (interface) part of the library, not the implementation: https://github.com/godotengine/godot/blob/master/thirdparty/minimp3/minimp3_ex.h#L1-L130

So it's basically a `.cpp` and `.h` in a trench-coat, and you need to pass those defines once to toggle the `.cpp` mode.

Reply:
> An alternative could be to add a `thirdparty/minimp3/minimp3_ex.cpp` that includes the header and set the implementation flags, so we don't need to do it in module code.

Did that with #78271.